### DISCUSSION TITLE
filtering by child collection

### DISCUSSION THREAD

**Discussion by dtau00** [appy-one/acebase#83] [Q&A] [answered]
Ive tried various things between db rules, queries, etc... not sure what the right approach is.  So i have a collection of groups.  within each group i have a collection of members (which contains a uid).  How do i get back all groups which a user is a member of?  I didnt see a way to filter by a collection of child's prop.  

I can probably get it by separating groups and members, but its hard to tell what the right approach is.

**Comment by appy-one** [comment, marked as answer]
Ok, so if I understand it correctly, this is your datamodel:

```json
{
   "groups": {
      "group1": {
          "members": {
             "member1": "user1uuid",
             "member2": "user2uuid",
             "member3": "user3uuid"
          }
      },
      "group2": {
         "members": {
            "member1": "user3uuid"
         }
      }
   }
}
```

And you'd want to get all groups that, say user `"user3uuid"` is a member of?

**Option 1:**

Convert your members collections to arrays:
```json
{
   "groups": {
      "group1": {
          "members": ["user1uuid", "user2uuid", "user3uuid"]
      },
      "group2": {
         "members": ["user3uuid"]
      }
   }
}
```
You can then create an array index (optional): `db.indexes.create('groups', 'members', { type: 'array' })` and query the groups with `db.query('groups').filter('members', 'contains', 'user3uuid').get()`

**Option 2 (more work):**

Convert the member strings to objects:
```json
{
   "groups": {
      "group1": {
          "members": {
             "member1": { "uuid": "user1uuid" },
             "member2": { "uuid": "user2uuid" },
             "member3": { "uuid": "user3uuid" }
          }
      },
      "group2": {
         "members": {
            "member1": { "uuid": "user3uuid" }
         }
      }
   }
}
```

You can then create the index (required): `db.indexes.create('groups/*/members', 'uuid')`, then:
```js
// Get references to matching member nodes with a query:
const refs = await db.query('groups/$groupId/members').filter('uuid', '==', 'user3uuid').getRefs();

// Extract groupId's from the paths
const groupIds = refs.map(ref => ref.vars.groupId);

// Fetch only the matched groups:
const snapshot = db.ref('groups').get({ include: groupIds });
// snapshot.val() contains { "group1": { ... }, "group2": { ... } }

// Iterate groups:
snapshot.forEach(snap => {
   console.log(`Found group ${snap.ref.key}: `, snap.val());
})
```

Note: no code above is tested, might contain typos...

Hope this helps!

**Reply by dtau00** [reply]
So i was able to get it working, but not exactly in the way its described.  Maybe this will help my understand and others that run into this.  I think there was a typo of between the index and the query path.  They both seem to need to match in order to work, or it will just return an empty result.  

   db.indexes.create('groups/$groupid/members', 'uid')
   db.query('groups/$groupId/members')...

or

   db.indexes.create('groups/*/members', 'uid')
   db.query('groups/*/members')...

next... const groupIds = refs.map(ref => ref.vars.groupid)
i could see that ref.vars is an object but couldn't peer much into it.  I tried various ways of extracting something out of it but couldnt get anything.  ref.vars.groupid, ref.vars[0], ref.vars['groupid'], ref.vars['id'], etc.. just kept getting undefined.  maybe because we are only getRef() that its not populating ref.vars?  So to get around this i was able to get the groupid by moving up the parent nodes. final code being:

  db.indexes.create('groups/*/members', 'uid')
  ...
  const refs = await db.query('groups/*/members').filter('uid', '==', user().uid).getRefs();
  const groupIds = refs.map(ref => ref.parent.parent.key)
  const snapshot = await db.ref('groups').get({ include: groupIds })
  let snaps = snapshot.val()
  for(let key in snaps)
      this.myGroups.push(snaps[key])

**Comment by dtau00** [comment]
awesome appreciate the quick response!  thanks for both examples, option two is what i needed (each member object has a bunch of properties with it).
